validator/resource_value_pattern_test.rego

```diff
@@ -0,0 +1,187 @@
+	{"is_required": true, "has_pattern": false, "has_field": false, "pattern": "^\\d+$", "value": "1234", "result_is_invalid": true},
+	{"is_required": true, "has_pattern": false, "has_field": false, "pattern": "^\\d+$", "value": "abcd", "result_is_invalid": true},
+	{"is_required": false, "has_pattern": true, "has_field": true, "pattern": "^\\d+$", "value": "1234", "result_is_invalid": false},
+	{"is_required": false, "has_pattern": true, "has_field": true, "pattern": "^\\d+$", "value": "abcd", "result_is_invalid": true},
+	{"is_required": false, "has_pattern": true, "has_field": false, "pattern": "^\\d+$", "value": "1234", "result_is_invalid": false},
+	{"is_required": false, "has_pattern": true, "has_field": false, "pattern": "^\\d+$", "value": "abcd", "result_is_invalid": false},
+	{"is_required": false, "has_pattern": false, "has_field": true, "pattern": "^\\d+$", "value": "1234", "result_is_invalid": false},
+	{"is_required": false, "has_pattern": false, "has_field": true, "pattern": "^\\d+$", "value": "abcd", "result_is_invalid": false},
+	{"is_required": false, "has_pattern": false, "has_field": false, "pattern": "^\\d+$", "value": "1234", "result_is_invalid": false},
+	{"is_required": false, "has_pattern": false, "has_field": false, "pattern": "^\\d+$", "value": "abcd", "result_is_invalid": false},
+]
+
+is_valid_bool(obj) = output {
+	is_valid(obj)
+	output := true
+}
+
+is_valid_bool(obj) = output {
+	not is_valid(obj)
+	output := false
+}
+
+all_is_valid_test_scenarios[[i, scenario]] {
+	t := validation_scenario_table[i]
+	d := {"is_required": t.is_required, "has_pattern": t.has_pattern, "has_field": t.has_field, "pattern": t.pattern, "value": t.value}
+	result := is_valid_bool(d)
+	result != t.result_is_invalid
+	scenario = {
+		"scenario": t,
+		"result": result,
+	}
+}
+
+test_is_valid_test_scenarios {
+	# ensure we are always returning a true / false from is_valid and not any unknowns
+	count(all_is_valid_test_scenarios) == count_of_test_scenarios
+}
+
+test_has_field_by_path {
+	doc := {
+		"labels": {
+			"f": false,
+			"id": "1234",
+			"t": true,
+			"tf_foo": "foo/project/google/v123",
+		},
+		"name": "myname",
+	}
+
+	has_field_by_path(doc, "name")
+	has_field_by_path(doc, "labels.id")
+	has_field_by_path(doc, "labels.f")
+	has_field_by_path(doc, "labels.t")
+
+	not has_field_by_path(doc, "unknown")
+	not has_field_by_path(doc, "labels.unknown")
+	not has_field_by_path(doc, "labels.f.unknown")
+	not has_field_by_path(doc, "labels.t.unknown")
+	not has_field_by_path(doc, "labels.tf_foo.unknown")
+}
+
+test_get_default_by_path {
+	doc := {
+		"labels": {
+			"f": false,
+			"id": "1234",
+			"t": true,
+			"tf_foo": "foo/project/google/v123",
+		},
+		"name": "myname",
+	}
+
+	get_default_by_path(doc, "name", "foo") == "myname"
+	get_default_by_path(doc, "labels.id", "foo") == "1234"
+	get_default_by_path(doc, "labels.t", false) == true
+	get_default_by_path(doc, "labels.f", true) == false
+
```

Comment:
add a comment here to denote that this block is verifying the default behavior? (or consider splitting this up to 2 tests)